### ISSUE TITLE
Nullable multipart parts causes parts to be buffered into memory

### ISSUE THREAD

**Issue by pacey** [micronaut-projects/micronaut-core#4016]
### Task List

- [x] Steps to reproduce provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Have a controller that accepts multipart form data in the form of `StreamingFileUpload` parts, where one of the parts is nullable.
2. Call the endpoint will all the parts, bar the nullable one, with the parts ordered by smallest to largest.

### Expected Behaviour

The controller should be called as soon as it has received data for each part.

### Actual Behaviour

The controller is only called after all the parts have been loaded into memory, assuming because it doesn't handle the nullable case.

### Environment Information

- **Operating System**: Mac OSX
- **Micronaut Version:** 2.0.1
- **JDK Version:** 14

### Example Application

- https://github.com/pacey/large-file-uploads

**Comment by jameskleeh** [comment]
@pacey I'm open to suggestions for improvement, but I don't know what we could do about this. The reason why is because we don't know if the nullable part is coming or not until the entire request is processed.

**Comment by pacey** [comment]
That's a good point @jameskleeh, I assume the client sends the parts sequentially rather than all of them in parallel.

I guess a workaround for our case is to use the `MultipartBody`, at least we can start handling the data as soon as possible without the need for Micronaut delaying the call to our controller.

**Comment by jameskleeh** [comment]
@pacey Right - the parts come in sequentially. We initially had a plan to call controllers empty publishers, but StreamingFileUpload at least requires 1 chunk to get the file name, content type, etc...

**Comment by archerz** [comment]
@jameskleeh Would it be possible to have `CompletedFileUpload`, or maybe another variant of it, streaming the files to disk and then calling the controller when the request is finished?

This would be similar to most other frameworks, I believe.

**Comment by jameskleeh** [comment]
@archerz Yes - see https://docs.micronaut.io/latest/guide/configurationreference.html#io.micronaut.http.server.HttpServerConfiguration$MultipartConfiguration

**Comment by archerz** [comment]
So it seems like we should have set `disk` to `true`, as opposed to just using the mixed mode...

**Comment by jameskleeh** [comment]
@archerz That usually is not the best option. That would flush all content to disk so that would cause a performance hit for small files that could just stay in memory

**Comment by archerz** [comment]
@pacey I think what we've done wrong is setting the limit too high, believing that it is the maximum multipart size we're allowing to be sent. It should be set to a reasonable number, and then anything higher gets buffered to disk with `mixed = true`'

**Comment by jameskleeh** [comment]
The `micronaut.server.multipart.threshold` option is what controls at what point the file will go to disk.

**Comment by jameskleeh** [comment]
Going to close this issue as a duplicate of https://github.com/micronaut-projects/micronaut-core/issues/2869
